org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,206 @@
+import java.io.IOException;
+import java.util.Objects;
+
+public final class DfsBlockCaffeineCache extends DfsBlockCache {
+
+    private static final Logger log = LoggerFactory.getLogger(DfsBlockCaffeineCache.class);
+
+    /** Pack files smaller than this size can be copied through the cache. */
+    private final long maxStreamThroughCache;
+
+    /**
+     * Suggested block size to read from pack files in bytes.
+     * <p>
+     * If a pack file does not have a native block size, this size will be used.
+     * <p>
+     * If a pack file has a native size, a whole multiple of the native size
+     * will be used until it matches this size.
+     * <p>
+     * The value for blockSize must be a power of 2 and no less than 512.
+     */
+    private final int blockSize;
+
+    /** Cache of pack files, indexed by description. */
+    private final Cache<DfsPackDescription, DfsPackFile> packFileCache;
+
+    /** Cache of Dfs blocks. */
+    private final Cache<DfsPackKeyWithPosition, Ref> dfsBlockCache;
+
+    public static void reconfigure(final DfsBlockCaffeineCacheConfig cacheConfig) {
+        DfsBlockCache.setInstance(new DfsBlockCaffeineCache(cacheConfig));
+    }
+
+    private DfsBlockCaffeineCache(DfsBlockCaffeineCacheConfig cacheConfig) {
+        maxStreamThroughCache = (long) (cacheConfig.getCacheMaximumSize() * cacheConfig.getStreamRatio());
+
+        blockSize = cacheConfig.getBlockSize();
+
+        // TODO: register the cache and expose the stats
+        // the estimated retained bytes is estimated based on the fields each object contains
+        packFileCache = Caffeine.newBuilder()
+                .removalListener((DfsPackDescription description, DfsPackFile packFile, RemovalCause cause) -> {
+                    if (packFile != null) {
+                        log.debug("PackFile {} is removed because it {}", packFile.getPackName(), cause);
+                        packFile.key.cachedSize.set(0);
+                        packFile.close();
+                    }})
+                .maximumWeight(cacheConfig.getCacheMaximumSize() / 2)
+                .weigher((DfsPackDescription description, DfsPackFile packFile) -> {
+                    // weight is static after creation and update, so here we are relying on dfsBlockCache's removal
+                    // listen to make sure the retained size of packFile won't exceed the given memory
+                    long estimatedSize = 2048 + blockSize;
+                    return estimatedSize > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) estimatedSize;
+                })
+                .recordStats()
+                .build();
+
+        dfsBlockCache = Caffeine.newBuilder()
+                .removalListener((DfsPackKeyWithPosition keyWithPosition, Ref ref, RemovalCause cause) -> ref = null)
+                .maximumWeight(cacheConfig.getCacheMaximumSize() / 2)
+                .weigher((DfsPackKeyWithPosition keyWithPosition, Ref ref) -> ref == null? 48 : 48 + ref.getSize())
```

Comment:
With this and the higher line, `48` and `2048` look like magical constants.  Can we make this clearer?